[Merged by Bors] - feat(FieldTheory/PurelyInseparable): introduce exponent

mistarro opened the pull request:
Define class of field extensions with exponent.
Show that if exponent exists then the extension is purely inseparable.
Define the exponent of an element.
Prove basic results.


---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

If you are moving or deleting declarations, please include these lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Moves:
- Vector.* -> List.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

mistarro commented:
I'm still not sure if I should rather use `[ExpChar K p]` instance-implicit parameter.

acmepjz commented on Mathlib/FieldTheory/PurelyInseparable/Exponent.lean:
```diff
@@ -0,0 +1,150 @@
+
+/-- 'Encoding' of an element `a : L` as the pair `⟨n, y⟩ : ℕ × K` with
+`minpoly K a = X ^ ringExpChar K ^ n - y`. -/
+noncomputable def elemEncode (a : L) : ℕ × K :=
+  let p := ringExpChar K
+  haveI : ExpChar K p := inferInstance
+  Classical.choose <| Prod.exists'.mpr (minpoly_eq_X_pow_sub_C K p a)
```
```suggestion
noncomputable def elemEncode (a : L) : ℕ × K :=
  Classical.choose <| Prod.exists'.mpr (minpoly_eq_X_pow_sub_C K (ringExpChar K) a)
```

(untested)

See also the comment I left at Zulip for a possibly better approach.

acmepjz commented on Mathlib/FieldTheory/PurelyInseparable/Exponent.lean:
```diff
@@ -0,0 +1,150 @@
+
+variable {L}
+
+open Classical in
+theorem exponent_def [HasExponent K L] (a : L) :
+    a ^ ringExpChar K ^ exponent K L ∈ (algebraMap K L).range :=
+  Nat.find_spec ‹HasExponent K L›.has_exponent' a
```
Maybe for this result (or another `exponent_def'`) you should use an input `p` together with `[ExpChar K p]`.

Also, you missed a result that if `e < exponent K L`, then there exists an `a` such that `a ^ ringExpChar K ^ e` is not in `K`.

mistarro replied:
Created versions with `ExpChar` for some theorems and added the result you suggested.

acmepjz reviewed (commented):
Thank you for your hard work! Unfortunately, I have no free time these days for reviewing the remaining parts yet. I think there can be some improvements.

mistarro commented:
@acmepjz Thanks for the review! Changed as suggested.

acmepjz commented on Mathlib/FieldTheory/PurelyInseparable/Exponent.lean:
```diff
@@ -0,0 +1,193 @@
+
+- `IsPurelyInseparable.HasExponent`: typeclass to assert a purely inseparable field extension
+  `L / K` has an exponent, that is a smallest natural number `e` such that
+  `a ^ ringExpChar K ^ e ∈ K`
+  for all `a ∈ L`.
```
Maybe
```suggestion
  `a ^ ringExpChar K ^ e ∈ K` for all `a ∈ L`.
```

acmepjz commented on Mathlib/FieldTheory/PurelyInseparable/Exponent.lean:
```diff
@@ -0,0 +1,193 @@
+    have h_elemexp_bound (a : L) : elemExponent K a ≤ e :=
+      Nat.le_log_of_pow_le (Nat.Prime.one_lt <| ringExpChar.eq K p ▸ hp)
+        (minpoly_natDegree_eq K a ▸ minpoly.natDegree_le a)
+    rw [RingHom.map_pow,
+      algebraMap_elemReduct_eq,
+      ← pow_mul, ← pow_add,
```
Maybe
```suggestion
    rw [RingHom.map_pow, algebraMap_elemReduct_eq, ← pow_mul, ← pow_add,
```

acmepjz commented on Mathlib/FieldTheory/PurelyInseparable/Exponent.lean:
```diff
@@ -0,0 +1,193 @@
+theorem elemExponent_eq_zero_of_mem_range {a : L} (h : a ∈ (algebraMap K L).range) :
+    elemExponent K a = 0 := by
+  apply (Nat.find_eq_zero _).mpr
+  simp
```
I think non-terminal `simp` is not allowed in mathlib code. Could you try replace it by the output of `simp?` which looks like `simp only [XXX]` ?

acmepjz reviewed (commented):
Thank you very much! I think now it's good to go besides some style suggestions.

acmepjz commented on Mathlib/FieldTheory/PurelyInseparable/Exponent.lean:
```diff
@@ -0,0 +1,190 @@
+  obtain ⟨y, hy⟩ := h
+  exact ⟨y, hy ▸ minpoly.eq_X_sub_C L y⟩
+
+open Classical in
```
This `open Classical` is not necessary? (untested)

acmepjz commented on Mathlib/FieldTheory/PurelyInseparable/Exponent.lean:
```diff
@@ -0,0 +1,190 @@
+open Classical in
+/-- The exponent of an element `a ∈ L` of a purely inseparable field extension `L / K`
+is the smallest natural number `e` such that `a ^ ringExpChar K ^ e ∈ K`. -/
+noncomputable def elemExponent (a : L) : ℕ :=
```
Do you think results` elemExponent_(def|min)[']` are useful?

mistarro replied:
Ok, will add them.

alreadydone replied:
I wonder if we should define the degree, sepDegree and insepDegree of an algebraic element; it seems you need the degree to define the absolute trace. You need to take `Nat.log` to get from `insepDegree` to `elemExponent` though.

acmepjz replied:
> I wonder if we should define the degree, sepDegree and insepDegree of an algebraic element; it seems you need the degree to define the absolute trace. You need to take `Nat.log` to get from `insepDegree` to `elemExponent` though.

I think it's reasonable, but maybe not in this PR. The definition of `elemExponent` could be changed to the smallest natural number `e` such that `a ^ ringExpChar K ^ e ∈ separableClosure K L`; in this case the `IsPurelyInseparable` assumption could be changed to `Algebra.IsAlgebraic`.

mistarro replied:
Note that it is just respective degree of the minimal polynomial. I think definitions stated in this PR are easier to work with, see for example the result `minpoly_eq`. Defining `elemExponent` just from degree would add complexity. And I need `elemReduct` for proceeding with the frobenius map.

mistarro commented:
I removed versions with `[ExpChar K p]` as they are always `ringExpChar K p ▸ ...` so it does not make sense to have them.

acmepjz commented:
> I removed versions with `[ExpChar K p]` as they are always `ringExpChar.eq K p ▸ ...` so it does not make sense to have them. I think API using `ringExpChar K` is clean.

That's fine. Would you like to add a remark to docstring explaining that how to convert `ringExpChar K` case to `[ExpChar K p]` case?

jcommelin commented on Mathlib/FieldTheory/PurelyInseparable/Exponent.lean:
```diff
@@ -0,0 +1,173 @@
+variable {L}
+
+open Classical in
+theorem exponent_def [HasExponent K L] (a : L) :
+    a ^ ringExpChar K ^ exponent K L ∈ (algebraMap K L).range :=
```
This lemma can be stated in terms of `[ExpChar K p]` to make it easier to apply.

jcommelin commented on Mathlib/FieldTheory/PurelyInseparable/Exponent.lean:
```diff
@@ -0,0 +1,173 @@
+
+/-- A predicate class on a ring extension saying that there is a natural number `e`
+such that `a ^ ringExpChar K ^ e ∈ K` for all `a ∈ L`. -/
+class HasExponent : Prop where
+  has_exponent : ∃ e, ∀ a, a ^ ringExpChar K ^ e ∈ (algebraMap K L).range
```
Please create a constructor for this class that *does* assume `[ExpChar K p]`.

acmepjz replied:
> Please create a constructor for this class that _does_ assume `[ExpChar K p]`.

... which I think is to add a `@[mk_iff]` before the `class HasExponent : Prop where`, and add a `hasExponent_iff'` which uses `[ExpChar K p]` (in this case I think `ringExpChar.eq K p ▸` is not convenient).

jcommelin replied:
I would prefer if the version that uses `ExpChar` is the main one, without the `'`.

jcommelin commented on Mathlib/FieldTheory/PurelyInseparable/Exponent.lean:
```diff
@@ -0,0 +1,173 @@
+variable {K}
+
+open Classical in
+theorem exponent_min [HasExponent K L] {e : ℕ} (h : e < exponent K L) :
+    ∃ a, a ^ ringExpChar K ^ e ∉ (algebraMap K L).range :=
```
Once again, this can be stated using `[ExpChar]`.

jcommelin commented on Mathlib/FieldTheory/PurelyInseparable/Exponent.lean:
```diff
@@ -0,0 +1,173 @@
+  Classical.choose <| Nat.find_spec <| minpoly_eq_X_pow_sub_C K (ringExpChar K) a
+
+open Classical in
+theorem minpoly_eq (a : L) :
+    minpoly K a = X ^ ringExpChar K ^ elemExponent K a - C (elemReduct K a) :=
```
Idem dito.

mistarro replied:
Initially I did like this for each result but all the proofs were just `ringExpChar.eq K p ▸ <corresponding result>`, so I removed and commented on it in the file docstring. Do you think it's worth duplicating (almost) everything if the usage in the context where `[ExpChar K p]` is available is only to do above substitution?

jcommelin replied:
Yes, because the substitution trick only works in "simple" contexts, like the API lemmas in this file.
Users of this file will eventually be thankful it is written with the generality of `[ExpChar K p]`.

(Also please don't resolve conversations before people have a chance to read and reply to your response.)

jcommelin reviewed (commented):
You correctly use `ringExpChar` in the definitions. But please use `[ExpChar K p]` in the api results about those definitions.

mistarro commented:
@jcommelin I hope I addressed your comments. I'm still curious what would be an example of usage of this new API where the trick with substitution using `ringExpChar.eq E p` would not work or would be inconvenient?

jcommelin commented:
As soon as you have a `K` and an `L` with the same characteristic, and you want to view them both as algebras over the same `ZMod p`.

jcommelin commented on Mathlib/FieldTheory/PurelyInseparable/Exponent.lean:
```diff
@@ -0,0 +1,195 @@
+
+/-- The degree of the minimal polynomial of an element `a ∈ L` equals
+`ringExpChar K ^ elemExponent K a`. -/
+theorem minpoly_natDegree_eq (a : L) :
+    (minpoly K a).natDegree = ringExpChar K ^ elemExponent K a := by
+  rw [minpoly_eq K a, natDegree_sub_C, natDegree_pow, natDegree_X, mul_one]
+
+theorem algebraMap_elemReduct_eq (a : L) :
+    algebraMap K L (elemReduct K a) = a ^ ringExpChar K ^ elemExponent K a := by
+  have := minpoly_eq K a ▸ minpoly.aeval K a
+  rwa [map_sub, aeval_C, map_pow, aeval_X, sub_eq_zero, eq_comm] at this
+
+theorem elemExponent_def (a : L) :
+    a ^ ringExpChar K ^ elemExponent K a ∈ (algebraMap K L).range :=
+  RingHom.mem_range.mpr <| ⟨_, algebraMap_elemReduct_eq K a⟩
+
+variable {K} in
+theorem elemExponent_le_of_pow_mem {a : L} {n : ℕ}
+    (h : a ^ ringExpChar K ^ n ∈ (algebraMap K L).range) : elemExponent K a ≤ n := by
+  let ⟨p, _⟩ := ExpChar.exists K
+  rcases ‹ExpChar K p› with _ | ⟨hp⟩
+  · exact elemExponent_eq_zero_of_charZero K a ▸ Nat.zero_le _
+  · obtain ⟨y, hy⟩ := RingHom.mem_range.mp <| h
+    let f := X ^ ringExpChar K ^ n - C y
+    have hf₁ : f.aeval a = 0 := by rwa [map_sub, aeval_C, aeval_X_pow, sub_eq_zero, eq_comm]
+    have hf₂ : f.Monic := monic_X_pow_sub_C y <| Nat.pos_iff_ne_zero.mp <| expChar_pow_pos K _ _
+    have hf₃ : f.natDegree = ringExpChar K ^ n := by
+      rw [natDegree_sub_C, natDegree_pow, natDegree_X, mul_one]
+    exact (Nat.pow_le_pow_iff_right <| Nat.Prime.one_lt hp).mp <|
+      ringExpChar.eq K p ▸ hf₃ ▸ minpoly_natDegree_eq K a ▸
+      natDegree_le_natDegree (minpoly.min K a hf₂ hf₁)
+
+theorem elemExponent_min (a : L) {n : ℕ} (h : n < elemExponent K a) :
+    a ^ ringExpChar K ^ n ∉ (algebraMap K L).range :=
```
Please also provide variants for these lemmas. After that, this seems ready for merging.

bors d+

mistarro commented:
bors r+
